### ISSUE TITLE
os.rename fails for t1w_ref_dimensions

### ISSUE THREAD

**Issue by soichih** [poldracklab/fmriprep#2190]
I am running fmriprep 1.5.10 docker container through singularity 3.5.2

```
+ singularity exec -e docker://poldracklab/fmriprep:1.5.10 /usr/local/miniconda/bin/fmriprep --notrack --resource-monitor --skip-bids-validation --md-only-boilerplate --stop-on-first-crash --use-plugin=multi_proc.yml --output-spaces fsaverage5 MNI152NLin2009cAsym --force-bbr --use-syn-sdc --fs-license-file=license.txt --skull-strip-template=NKI --work-dir=fmripworkdir --participant_label=sub-TTTEMPSUB fmripinput fmripout participant
```

When fmriprep tries to rename `_0x65e1d82ccb813ce88af870b9b16d815c_unfinished.json'` to `_0x65e1d82ccb813ce88af870b9b16d815c` it fails with this error message.

```
You are using fMRIPrep-1.5.10, and a newer version of fMRIPrep is available: 20.0.7.
Please check out our documentation about how and when to upgrade:
https://fmriprep.readthedocs.io/en/latest/faq.html#upgrading
/usr/local/miniconda/lib/python3.7/site-packages/fmriprep/cli/run.py:547: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  plugin_settings = loadyml(f)
Downloading https://templateflow.s3.amazonaws.com/tpl-NKI/tpl-NKI_res-01_T1w.nii.gz

  0%|          | 0.00/42.1k [00:00<?, ?B/s]
  0%|          | 67.0/42.1k [00:00<01:02, 670B/s]
  1%|          | 247/42.1k [00:00<00:50, 825B/s] 
  2%|▏         | 876/42.1k [00:00<00:36, 1.12kB/s]
  8%|▊         | 3.46k/42.1k [00:00<00:24, 1.57kB/s]
 17%|█▋        | 7.10k/42.1k [00:00<00:15, 2.20kB/s]
 23%|██▎       | 9.81k/42.1k [00:00<00:10, 3.03kB/s]
 33%|███▎      | 13.8k/42.1k [00:00<00:06, 4.17kB/s]
 41%|████      | 17.3k/42.1k [00:00<00:04, 5.66kB/s]
 51%|█████     | 21.4k/42.1k [00:00<00:02, 7.64kB/s]
 60%|██████    | 25.3k/42.1k [00:01<00:01, 10.1kB/s]
 69%|██████▉   | 29.1k/42.1k [00:01<00:01, 12.9kB/s]
 78%|███████▊  | 32.9k/42.1k [00:01<00:00, 16.1kB/s]
 88%|████████▊ | 36.9k/42.1k [00:01<00:00, 19.6kB/s]
 97%|█████████▋| 40.9k/42.1k [00:01<00:00, 23.1kB/s]
42.1kB [00:01, 28.9kB/s]                            
Downloading https://templateflow.s3.amazonaws.com/tpl-NKI/tpl-NKI_res-01_label-brain_probseg.nii.gz

  0%|          | 0.00/2.32k [00:00<?, ?B/s]
  3%|▎         | 77.0/2.32k [00:00<00:04, 546B/s]
 20%|██        | 467/2.32k [00:00<00:02, 735B/s] 
 82%|████████▏ | 1.91k/2.32k [00:00<00:00, 1.03kB/s]
2.32kB [00:00, 6.51kB/s]                            
Downloading https://templateflow.s3.amazonaws.com/tpl-NKI/tpl-NKI_res-01_desc-BrainCerebellumExtraction_mask.nii.gz

  0%|          | 0.00/201 [00:00<?, ?B/s]
 51%|█████     | 102/201 [00:00<00:00, 1.00kB/s]
202B [00:00, 1.29kB/s]                          
Downloading https://templateflow.s3.amazonaws.com/tpl-MNI152NLin2009cAsym/tpl-MNI152NLin2009cAsym_res-02_desc-fMRIPrep_boldref.nii.gz

  0%|          | 0.00/1.71k [00:00<?, ?B/s]
  4%|▍         | 68.0/1.71k [00:00<00:02, 677B/s]
 16%|█▌        | 266/1.71k [00:00<00:01, 843B/s] 
 60%|█████▉    | 1.02k/1.71k [00:00<00:00, 1.15kB/s]
1.71kB [00:00, 5.32kB/s]                            
Downloading https://templateflow.s3.amazonaws.com/tpl-MNI152NLin2009cAsym/tpl-MNI152NLin2009cAsym_res-02_desc-brain_mask.nii.gz

  0%|          | 0.00/28.0 [00:00<?, ?B/s]
29.0B [00:00, 550B/s]                     
Downloading https://templateflow.s3.amazonaws.com/tpl-MNI152NLin2009cAsym/tpl-MNI152NLin2009cAsym_res-01_desc-carpet_dseg.nii.gz

  0%|          | 0.00/440 [00:00<?, ?B/s]
 27%|██▋       | 119/440 [00:00<00:00, 1.19kB/s]
 93%|█████████▎| 410/440 [00:00<00:00, 1.44kB/s]
441B [00:00, 2.16kB/s]                          
Downloading https://templateflow.s3.amazonaws.com/tpl-MNI152NLin2009cAsym/tpl-MNI152NLin2009cAsym_res-01_T1w.nii.gz

  0%|          | 0.00/13.4k [00:00<?, ?B/s]
  1%|▏         | 169/13.4k [00:00<00:07, 1.69kB/s]
  3%|▎         | 356/13.4k [00:00<00:07, 1.74kB/s]
  8%|▊         | 1.02k/13.4k [00:00<00:05, 2.23kB/s]
 28%|██▊       | 3.70k/13.4k [00:00<00:03, 3.06kB/s]
 55%|█████▌    | 7.38k/13.4k [00:00<00:01, 4.22kB/s]
 78%|███████▊  | 10.4k/13.4k [00:00<00:00, 5.68kB/s]
13.4kB [00:00, 19.6kB/s]                            
Downloading https://templateflow.s3.amazonaws.com/tpl-MNI152NLin2009cAsym/tpl-MNI152NLin2009cAsym_res-01_desc-brain_mask.nii.gz

  0%|          | 0.00/155 [00:00<?, ?B/s]
 55%|█████▍    | 85.0/155 [00:00<00:00, 845B/s]
156B [00:00, 1.03kB/s]                         
fMRIPrep failed: Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/shutil.py", line 557, in move
    os.rename(src, real_dst)
FileNotFoundError: [Errno 2] No such file or directory: '/export/prod/5ecbc45ffc80c1414d1e4c3a/5ee7cd51d1b4c1e3eb234fa6/fmripworkdir/fmriprep_wf/single_subject_TTTEMPSUB_wf/anat_preproc_wf/anat_template_wf/t1w_ref_dimensions/_0x65e1d82ccb813ce88af870b9b16d815c_unfinished.json' -> '/export/prod/5ecbc45ffc80c1414d1e4c3a/5ee7cd51d1b4c1e3eb234fa6/fmripworkdir/fmriprep_wf/single_subject_TTTEMPSUB_wf/anat_preproc_wf/anat_template_wf/t1w_ref_dimensions/_0x65e1d82ccb813ce88af870b9b16d815c.json'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/plugins/legacymultiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 531, in run
    shutil.move(hashfile_unfinished, hashfile_unfinished.replace("_unfinished", ""))
  File "/usr/local/miniconda/lib/python3.7/shutil.py", line 571, in move
    copy_function(src, real_dst)
  File "/usr/local/miniconda/lib/python3.7/shutil.py", line 257, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/local/miniconda/lib/python3.7/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/export/prod/5ecbc45ffc80c1414d1e4c3a/5ee7cd51d1b4c1e3eb234fa6/fmripworkdir/fmriprep_wf/single_subject_TTTEMPSUB_wf/anat_preproc_wf/anat_template_wf/t1w_ref_dimensions/_0x65e1d82ccb813ce88af870b9b16d815c_unfinished.json'

fMRIPrep could not find the markdown version of the citation boilerplate (/export/prod/5ecbc45ffc80c1414d1e4c3a/5ee7cd51d1b4c1e3eb234fa6/fmripout/fmriprep/logs/CITATION.md). HTML and LaTeX versions of it will not be available
Preprocessing did not finish successfully. Errors occurred while processing data from participants: TTTEMPSUB (1). Check the HTML reports for details.

real	6m38.263s
user	0m21.004s
sys	0m10.214s
```

Strangely, I see the .json file already renamed in the directory.

```
user@slurm7b:/export/prod/5ecbc45ffc80c1414d1e4c3a/5ee7cd51d1b4c1e3eb234fa6/fmripworkdir/fmriprep_wf/single_subject_TTTEMPSUB_wf/anat_preproc_wf/anat_template_wf/t1w_ref_dimensions$ ls
_0x65e1d82ccb813ce88af870b9b16d815c.json  _node.pklz  report.html
_inputs.pklz                              _report     result_t1w_ref_dimensions.pklz
```

I don't know why this happens, but it seems that fmriprep is trying to rename a file that's already renamed. Or maybe trying to rename the same file twice? I am not sure..

Here is the content of `_0x65e1d82ccb813ce88af870b9b16d815c.json` 

```
[
    [
        "max_scale",
        "3.0000000000"
    ],
    [
        "t1w_list",
        [
            [
                "/export/prod/5ecbc45ffc80c1414d1e4c3a/5ee7cd51d1b4c1e3eb234fa6/fmripinput/sub-TTTEMPSUB/ses-SSSES/anat/sub-TTTEMPSUB_ses-SSSES_acq-mprage_run-1_T1w.nii.gz",
                "c11789fe99638fc2636226e2144bb5b5"
            ]
        ]
    ],
    [
        "needed_outputs",
        [
            "out_report",
            "t1w_valid_list",
            "target_shape",
            "target_zooms"
        ]
    ]
]
```

**Comment by oesteban** [comment]
I typically see that when 2 processes are running the same interface almost simultaneously in parallel. One of them finishes first, renames the file and the other fails in its attempt.

**Comment by oesteban** [comment]
Any updates on this? @soichih, did you find the root of the problem?

**Comment by soichih** [comment]
@oesteban Sorry, I haven't really revisited this since I reported it. I will put some effort into this soon.

**Comment by giulia-berto** [comment]
Hi @oesteban, @soichih asked me to do some tests to reproduce this issue. I tried exactly what he did, i.e. the same command with 1.5.10 docker container through singularity 3.5.2, but I did not encounter this error. The preprocessing ended up successfully I guess ('fMRIPrep finished successfully!' did not appear though). I then ran the command again with the updated container 20.1.2 and it ran successfully. Also the apps on brainlife seem to work with no issues.
Maybe as you said, it was just an issue of concurrent jobs working on the same working directory!?
